### PR TITLE
[Merged by Bors] - expose extras from gltf nodes

### LINKED ISSUE #2153: gltf extras component
## What problem does this solve or what need does it fill?

In my game i would like to store a few variables in the gltf "extras" field in blender and use them to set physics properties on the created entities.

## What solution would you like?

Create a struct to store extras data, something like:
```rust
pub struct GltfExtras(pub serde_json::Map<String, serde_json::Value>)
```
and add it when creating entities in the gltf loader. I'm not sure as to what the best type to store the data would be,
according to the [gltf spec](https://github.com/KhronosGroup/glTF/tree/master/specification/2.0#extras) they should be represented as an object whenever possible, so either a hashmap or something similar.

## What alternative(s) have you considered?

I don't think there is another way to do this, but i would like suggestions if there are.

### PR DESCRIPTION
**Pull request by mockersf**
fixes #2153 

expose the `extras` field value as a string

### CODE REVIEW

File: crates/bevy_gltf/src/lib.rs [hunk 1]
```diff
@@ -58,3 +60,9 @@ pub struct GltfPrimitive {
     pub mesh: Handle<Mesh>,
     pub material: Option<Handle<StandardMaterial>>,
 }
+
+#[derive(Debug, Reflect, Default)]
```
Comment (NathanSWard) [inline crates/bevy_gltf/src/lib.rs:64]:
```suggestion
#[derive(Clone, Debug, Default, Reflect)]
```

Comment (NathanSWard) [reply crates/bevy_gltf/src/lib.rs:64]: I don't see why `Clone` shouldn't be supported (unless there is a specific reason) :)

**Review by NathanSWard — APPROVED** [review]

**Review by Frizi — APPROVED** [review]

**Comment by cart** [comment]
Lazy questions:

* How often are "extras" defined in gltf files
* Is `gltf_node.extras()` None when gltf extras aren't defined in the file? I'm assuming the answer is yes. I just want to sanity check that we aren't allocating a bunch of empty strings per-node/entity each time we spawn a GLTF scene.

**Comment by NathanSWard** [comment]
> I just want to sanity check that we aren't allocating a bunch of empty strings per-node/entity each time we spawn a GLTF scene.

Depending on this, we could always put this behind a feature flag (disabled by default) to avoid the extra allocations. 🤷

**Comment by mockersf** [comment]
> * How often are "extras" defined in gltf files

I don't know, I didn't saw yet in a model I used but it seems a nice place to put some metadata that I put in code until now

> * Is `gltf_node.extras()` None when gltf extras aren't defined in the file? I'm assuming the answer is yes. I just want to sanity check that we aren't allocating a bunch of empty strings per-node/entity each time we spawn a GLTF scene.

Yup, just checked with FlightHelmet, no extras added

**Comment by cart** [comment]
Just checked out a complicated gltf from my game High Hat and extras weren't defined.

**Comment by cart** [comment]
It looks like blender doesn't include extras by default. I think validating that other popular 3d packages dont include extras unless the user defines them is probably the "complete" solution here, but I understand that this hard to verify. Can we just "randomly" sample a few free gltf assets from some "model sharing" website and see if they include the field?

**Comment by mockersf** [comment]
I didn't find files with those in my random sample of files on my drive that I collected here and there (kenney's assets and a few other free downloads), and it seems only babylon.js supports this out of the box (but that's easy as json support is native in js...)

On the other hand I don't think it add much code or complexity, but it does add an extra component with a string that the user will have to deserialize... not the nicest UX

**Comment by quinnyo** [comment]
I would like to see this feature in bevy. I often find situations/projects where I want to be able to specify some metadata in a 3D package (Blender) and have it exported with the scene/model. GLTF extras is a solution to this problem.

Some points:

- extras can be included with many(/most/all?) gltf object types -- not just nodes
- I think it would be acceptable to only support / expect the extras data type to be a json object
- the gltf spec recommends using the object data type as a best practice
- Blender specific:
  - Blender exports 'Custom Properties' as extras if an option is enabled when exporting
  - the property values can be complex (~json objects) -- e.g. a value of `[1, 2, 3]` in Blender will be exported as an array of integers
  - Custom Properties in the Scene, Object, Object Data (Mesh), and Material are exported to their GLTF counterparts
  - https://docs.blender.org/manual/en/latest/addons/import_export/scene_gltf2.html#custom-properties

**Comment by nicopap** [comment]
I can confirm this could be very helpful. Currently working around the limitation with a trait that just pushes components on nodes with specific names. [my skin hook impl](https://gist.github.com/nicopap/a237bde5118ef38084ce51e8b759fb9f)

**Comment by cart** [comment]
We've been steadily increasing the number of "per entity string allocations" when instancing GLTF scene. I think we should prioritize "proper string cows" in the very near future to keep this behavior in check.

**Pull request closed without merging**